### ISSUE TITLE
BodyComponent prepareCanvas Method Throws UnimplementedError

### ISSUE THREAD

**Issue by mjordan56** [flame-engine/flame_box2d#42]
Build 0.7.2--releasecandidate.12 of the flame_forge package throws an UnimplementedError from the abstract class BodyComponent. This bug was introduced in the "Update to rc 12" commit `4058eaca79ba1b648120bcfe839624e9aa5f0110`. The bug is in line 58 of body_component.dart:

```
   if (_transform.m14 != body.position.x || 
       _transform.m24 != body.position.y ||
       _lastAngle != angle) { 
     _transform.setIdentity();  
     _transform.scale(1, -1);          // <- THIS IS THE BUG CAUSING THE UNIMPLEMENTEDERROR
     _transform.translate2(body.position); 
     _transform.rotateZ(angle);
     _lastAngle = angle;
   }
```

The UnimplementedError is being thrown because the Matrix4 scale method doesn't support integer values for the x parameter. The method signature is:

```
   void scale(dynamic x, [double? y, double? z])
```

The x parameter must be a Vector3, Vector4, or a double otherwise, an UnimplementedError is thrown.

**Comment by spydon** [comment]
Hi,
Can you please post the whole stack trace?
Dart can convert integers to double if they are const, so 1 represents `x` here and `-1` represents `y`.
Which platform are you running this on? And are you running in release- or debug-mode?

**Comment by mjordan56** [comment]
Hi @spydon, while Dart can convert integers to double I believe the issue here is that x is not declared a double. It's declared as dynamic hence x can be an integer value and it will not be cast as a double. It will preserve the data type of x as an integer. The Matrix4 scale method tests the type of the x parameter. If it isn't a Vector3, Vector4 or a double, the method throws an UnimplementedError. Here's the Matrix4 scale method that's throwing the error:

```
/// Scale this matrix by a [Vector3], [Vector4], or x,y,z
  void scale(dynamic x, [double? y, double? z]) {
    double sx;
    double sy;
    double sz;
    final sw = x is Vector4 ? x.w : 1.0;
    if (x is Vector3) {
      sx = x.x;
      sy = x.y;
      sz = x.z;
    } else if (x is Vector4) {
      sx = x.x;
      sy = x.y;
      sz = x.z;
    } else if (x is double) {
      sx = x;
      sy = y ?? x;
      sz = z ?? x;
    } else {
      throw UnimplementedError();
    }
    _m4storage[0] *= sx;
    _m4storage[1] *= sx;
    _m4storage[2] *= sx;
    _m4storage[3] *= sx;
    _m4storage[4] *= sy;
    _m4storage[5] *= sy;
    _m4storage[6] *= sy;
    _m4storage[7] *= sy;
    _m4storage[8] *= sz;
    _m4storage[9] *= sz;
    _m4storage[10] *= sz;
    _m4storage[11] *= sz;
    _m4storage[12] *= sw;
    _m4storage[13] *= sw;
    _m4storage[14] *= sw;
    _m4storage[15] *= sw;
  }
```

This is from the pub.dartlang.org > vector_math-2.1.0 > lib > src > vector_math_64 > matrix4.dart file. I'm running on a MacBook Pro, Flutter stable channel, 2.2.1, Dart version 2.13.1. The simple fix is to change the x parameter from 1 to 1.0.

**Comment by spydon** [comment]
I'm guessing you are running on android or iphone and that the cast works differenly on web (since js numbers...), because the examples work on web. I'll publish a fix with them as doubles when I get home. :)

**Comment by mjordan56** [comment]
@spydon you are correct, I'm running on both Android and iOS. The bug happens on both mobile platforms. Thanks for the quick reply.

BTW, thank you for all your work on forge2d and flame_forge2d!

**Comment by spydon** [comment]
Now it is fixed on `0.7.3-releasecandidate.12`, thanks for reporting the issue!
